### PR TITLE
Header Menu

### PR DESCRIPTION
**Pull request by aboodswiti**

### CODE REVIEW

**Review by bayanseder — APPROVED** [review]

File: client/src/components/Common/Header/Header.js [hunk 1]
```diff
@@ -1,48 +1,130 @@
...
-    const { history } = this.props
-    axios.get('/api/check').then(({ data }) => {
-      const { success, email, userId, userName } = data
+  componentDidMount() {
+    document.addEventListener("mousedown", this.handleClickOutside);
+  }
+  componentWillUnmount() {
+    document.removeEventListener("mousedown", this.handleClickOutside);
+  }
+  handleClickOutside = (event) => {
```
Comment (yaqootturman) [inline client/src/components/Common/Header/Header.js:39]: document this

Comment (yaqootturman) [reply client/src/components/Common/Header/Header.js:39]: In the Booking form page there is a 'logout' word from the previous version!

**Review by yaqootturman — CHANGES_REQUESTED** [review]

**Review by yaqootturman — APPROVED** [review]

File: client/src/components/Common/Header/Header.js [hunk 2]
```diff
@@ -1,48 +1,134 @@
...
+import { withRouter } from "react-router";
+import axios from "axios";
+import Login from "../../Layout/Login/Login";
 
 class Header extends Component {
-  constructor (props) {
-    super(props)
+  constructor(props) {
+    super(props);
+    this.container = React.createRef();
```
Comment (11janci) [inline client/src/components/Common/Header/Header.js:18]: it'd be good if the name indicated that it is a ref -> `containerRef`

File: client/src/components/Common/Header/Header.js [hunk 3]
```diff
@@ -1,48 +1,134 @@
...
+    }
+  };
+  handleButtonClick = () => {
+    axios.get("/api/check").then(({ data }) => {
+      const { success } = data;
+
+      this.setState((state) => {
+        return {
+          open: !state.open,
+          success: success,
```
Comment (11janci) [inline client/src/components/Common/Header/Header.js:56]: better name for `success` would be `isLoggedIn`

**Review by 11janci — COMMENTED** [review]
The overall mechanism can be simplified. You have the `handleButtonClick()` method that calls the `check` API and sets the `success` variable, based on which you display either `Home` or `Login`. But then, once the user clicks on Home/Login, you call the `check` API again to determine where to redirect. That's inefficient and unnecessary. It is true that the session might expire between the initial check and the redirect, however that should be handled by whatever component it is redirected to - any API calls in that component should fail on 401/403 and the component should in that case redirect to Login. As I mentioned before, the `check` API is really unfortunate, however it's too late to refactor that now.

Related to this is another problem, and that is that you do all of this in button's onclick handlers. So when a user clicks on the menu, you call the API, which could take seconds to respond, rendering the application unresponsive. Instead, the `check` API should be called in the `componentDidMount()` method once and then follow the pattern explained above.

**Pull request merged**